README.md
# ethereum-block-extractor
go-ethereum block information extracting module using Golang.


data/history

eth.blockNumber
admin.nodeInfo.enode
eth.blockNumber
admin.nodeInfo.enode
blockinfo
blockInfo
var blockNumber;
var blockInfo;
web3.eth.getBlockNumber(function(e, r) { blockNumber = r; });
blockNumber;
web3.eth.getBlock(blockNumber, function(e, r) { blockInfo = r; });
blockInfo
var blockNumber;
var blockInfo;
web3.eth.getBlockNumber(function(e, r) { blockNumber = r; });
blockNumber;
web3.eth.getBlock(blockNumber, function(e, r) { blockInfo = r; });
blockInfo
clear
var blockNumber;
var blockInfo;
web3.eth.getBlockNumber(function(e, r) { blockNumber = r; });
blockNumber;
web3.eth.getBlock(blockNumber, function(e, r) { blockInfo = r; });
blockInfo
admin.nodeInfo.enode
block
block.Number()
blockInfo
admin.nodeInfo.enode
eth.blockNumber
logs
eth.blockNumber
var blockNumber;
var blockInfo;
web3.eth.getBlockNumber(function(e, r) { blockNumber = r; });
var blockNumber;
var blockInfo;
web3.eth.getBlockNumber(function(e, r) { blockNumber = r; });
blockinfo
blockInfo
web3.eth.getBlock(blockNumber, function(e, r) { blockInfo = r; });
eth.block
eth.blockNumber
eth.getBlock(5607)
eth.getBlock(5608)
eth.getBlock(5662)
eth.getBlock(5661)
eth.getBlock(5709)

run_test.sh
#!/bin/bash
# INLab, Dongguk Univ.

geth --datadir /home/inlab/inlab-ethereum-hp2p/data/ init genesis.json
geth --networkid 15 --datadir data --port 30331 --mine --miner.threads=1 --miner.etherbase=0x0000000000000000000000000000000000000000 console 2>> /home/inlab/inlab-ethereum-hp2p/data/client1.log


extractor/test.go
package main

import (
	"fmt"
	"io"
	"net"

	extractor "github.com/docbull/ethereum-block-extractor/extractor/proto"
	proto "github.com/golang/protobuf/proto"
)

func ReceiveBlockInfo(conn net.Conn) {
	data := make([]byte, 0)
	buf := make([]byte, 1024)
	length := 0

	for {
		n, err := conn.Read(buf)
		if err != nil {
			if err != io.EOF {
				fmt.Println(err)
				return
			}
			break
		}

		data = append(data, buf[:n]...)
		length += n
	}

	block := &extractor.BlockInfo{}
	err := proto.Unmarshal(data[:length], block)
	if err != nil {
		fmt.Println(err)
		conn.Write([]byte("error"))
		return
	}

	fmt.Println("difficulty:", block.Difficulty)
	fmt.Println("extraData:", block.Extra)
	fmt.Println("gasLimit:", block.GasLimit)
	fmt.Println("gasUsed:", block.GasUsed)
	fmt.Println("hash:", block.TxHash)
	fmt.Println("logsBloom:", block.Bloom)
	fmt.Println("miner:", block.Coinbase)
	fmt.Println("mixHash:", block.MixDigest)
	fmt.Println("nonce:", block.Nonce)
	fmt.Println("number:", block.Number)
	fmt.Println("parentHash:", block.ParentHash)
	fmt.Println("receiptsRoot:", block.ReceiptHash)
	fmt.Println("sha3Uncles:", block.UncleHash)
	fmt.Println("stateRoot:", block.Root)
	fmt.Println("timestamp:", block.Time)
	fmt.Println("transactionRoot:", block.TxHash)
}

func Start2Listen() {
	lis, err := net.Listen("tcp", ":4242")
	if err != nil {
		fmt.Println(err)
		return
	}
	defer lis.Close()

	fmt.Println("Ethereum Block Extractor ...")
	for {
		conn, err := lis.Accept()
		fmt.Println("SOME PEER CONNECTED!")
		if err != nil {
			fmt.Println(err)
			return
		}

		go ReceiveBlockInfo(conn)
	}
}

func main() {
	Start2Listen()
}


extractor/proto/ethaudit.pb.go
// Copyright INLab, Dongguk Univ.

// Code generated by protoc-gen-go. DO NOT EDIT.
// versions:
// 	protoc-gen-go v1.26.0
// 	protoc        v3.6.1
// source: proto/ethaudit.proto

package proto

import (
	protoreflect "google.golang.org/protobuf/reflect/protoreflect"
	protoimpl "google.golang.org/protobuf/runtime/protoimpl"
	reflect "reflect"
	sync "sync"
)

const (
	// Verify that this generated code is sufficiently up-to-date.
	_ = protoimpl.EnforceVersion(20 - protoimpl.MinVersion)
	// Verify that runtime/protoimpl is sufficiently up-to-date.
	_ = protoimpl.EnforceVersion(protoimpl.MaxVersion - 20)
)

type StatusCode int32

const (
	StatusCode_Unknwon StatusCode = 0
	StatusCode_Ok      StatusCode = 1
	StatusCode_Failed  StatusCode = 2
)

// Enum value maps for StatusCode.
var (
	StatusCode_name = map[int32]string{
		0: "Unknwon",
		1: "Ok",
		2: "Failed",
	}
	StatusCode_value = map[string]int32{
		"Unknwon": 0,
		"Ok":      1,
		"Failed":  2,
	}
)

func (x StatusCode) Enum() *StatusCode {
	p := new(StatusCode)
	*p = x
	return p
}

func (x StatusCode) String() string {
	return protoimpl.X.EnumStringOf(x.Descriptor(), protoreflect.EnumNumber(x))
}

func (StatusCode) Descriptor() protoreflect.EnumDescriptor {
	return file_proto_ethaudit_proto_enumTypes[0].Descriptor()
}

func (StatusCode) Type() protoreflect.EnumType {
	return &file_proto_ethaudit_proto_enumTypes[0]
}

func (x StatusCode) Number() protoreflect.EnumNumber {
	return protoreflect.EnumNumber(x)
}

// Deprecated: Use StatusCode.Descriptor instead.
func (StatusCode) EnumDescriptor() ([]byte, []int) {
	return file_proto_ethaudit_proto_rawDescGZIP(), []int{0}
}

type BlockInfo struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	ParentHash  []byte `protobuf:"bytes,1,opt,name=ParentHash,proto3" json:"ParentHash,omitempty"`
	UncleHash   []byte `protobuf:"bytes,2,opt,name=UncleHash,proto3" json:"UncleHash,omitempty"`
	Coinbase    []byte `protobuf:"bytes,3,opt,name=Coinbase,proto3" json:"Coinbase,omitempty"`
	Root        []byte `protobuf:"bytes,4,opt,name=Root,proto3" json:"Root,omitempty"`
	TxHash      []byte `protobuf:"bytes,5,opt,name=TxHash,proto3" json:"TxHash,omitempty"`
	ReceiptHash []byte `protobuf:"bytes,6,opt,name=ReceiptHash,proto3" json:"ReceiptHash,omitempty"`
	Bloom       []byte `protobuf:"bytes,7,opt,name=Bloom,proto3" json:"Bloom,omitempty"`
	Difficulty  uint64 `protobuf:"varint,8,opt,name=Difficulty,proto3" json:"Difficulty,omitempty"`
	Number      uint64 `protobuf:"varint,9,opt,name=Number,proto3" json:"Number,omitempty"`
	GasLimit    uint64 `protobuf:"varint,10,opt,name=GasLimit,proto3" json:"GasLimit,omitempty"`
	GasUsed     uint64 `protobuf:"varint,11,opt,name=GasUsed,proto3" json:"GasUsed,omitempty"`
	Time        uint64 `protobuf:"varint,12,opt,name=Time,proto3" json:"Time,omitempty"`
	Extra       []byte `protobuf:"bytes,13,opt,name=Extra,proto3" json:"Extra,omitempty"`
	MixDigest   []byte `protobuf:"bytes,14,opt,name=MixDigest,proto3" json:"MixDigest,omitempty"`
	Nonce       uint64 `protobuf:"varint,15,opt,name=Nonce,proto3" json:"Nonce,omitempty"`
	BaseFee     uint64 `protobuf:"varint,16,opt,name=BaseFee,proto3" json:"BaseFee,omitempty"`
}

func (x *BlockInfo) Reset() {
	*x = BlockInfo{}
	if protoimpl.UnsafeEnabled {
		mi := &file_proto_ethaudit_proto_msgTypes[0]
		ms := protoimpl.X.MessageStateOf(protoimpl.Pointer(x))
		ms.StoreMessageInfo(mi)
	}
}

func (x *BlockInfo) String() string {
	return protoimpl.X.MessageStringOf(x)
}

func (*BlockInfo) ProtoMessage() {}

func (x *BlockInfo) ProtoReflect() protoreflect.Message {
	mi := &file_proto_ethaudit_proto_msgTypes[0]
	if protoimpl.UnsafeEnabled && x != nil {
		ms := protoimpl.X.MessageStateOf(protoimpl.Pointer(x))
		if ms.LoadMessageInfo() == nil {
			ms.StoreMessageInfo(mi)
		}
		return ms
	}
	return mi.MessageOf(x)
}

// Deprecated: Use BlockInfo.ProtoReflect.Descriptor instead.
func (*BlockInfo) Descriptor() ([]byte, []int) {
	return file_proto_ethaudit_proto_rawDescGZIP(), []int{0}
}

func (x *BlockInfo) GetParentHash() []byte {
	if x != nil {
		return x.ParentHash
	}
	return nil
}

func (x *BlockInfo) GetUncleHash() []byte {
	if x != nil {
		return x.UncleHash
	}
	return nil
}

func (x *BlockInfo) GetCoinbase() []byte {
	if x != nil {
		return x.Coinbase
	}
	return nil
}

func (x *BlockInfo) GetRoot() []byte {
	if x != nil {
		return x.Root
	}
	return nil
}

func (x *BlockInfo) GetTxHash() []byte {
	if x != nil {
		return x.TxHash
	}
	return nil
}

func (x *BlockInfo) GetReceiptHash() []byte {
	if x != nil {
		return x.ReceiptHash
	}
	return nil
}

func (x *BlockInfo) GetBloom() []byte {
	if x != nil {
		return x.Bloom
	}
	return nil
}

func (x *BlockInfo) GetDifficulty() uint64 {
	if x != nil {
		return x.Difficulty
	}
	return 0
}

func (x *BlockInfo) GetNumber() uint64 {
	if x != nil {
		return x.Number
	}
	return 0
}

func (x *BlockInfo) GetGasLimit() uint64 {
	if x != nil {
		return x.GasLimit
	}
	return 0
}

func (x *BlockInfo) GetGasUsed() uint64 {
	if x != nil {
		return x.GasUsed
	}
	return 0
}

func (x *BlockInfo) GetTime() uint64 {
	if x != nil {
		return x.Time
	}
	return 0
}

func (x *BlockInfo) GetExtra() []byte {
	if x != nil {
		return x.Extra
	}
	return nil
}

func (x *BlockInfo) GetMixDigest() []byte {
	if x != nil {
		return x.MixDigest
	}
	return nil
}

func (x *BlockInfo) GetNonce() uint64 {
	if x != nil {
		return x.Nonce
	}
	return 0
}

func (x *BlockInfo) GetBaseFee() uint64 {
	if x != nil {
		return x.BaseFee
	}
	return 0
}

type Status struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	Code StatusCode `protobuf:"varint,1,opt,name=Code,proto3,enum=ethaudit.StatusCode" json:"Code,omitempty"`
}

func (x *Status) Reset() {
	*x = Status{}
	if protoimpl.UnsafeEnabled {
		mi := &file_proto_ethaudit_proto_msgTypes[1]
		ms := protoimpl.X.MessageStateOf(protoimpl.Pointer(x))
		ms.StoreMessageInfo(mi)
	}
}

func (x *Status) String() string {
	return protoimpl.X.MessageStringOf(x)
}

func (*Status) ProtoMessage() {}

func (x *Status) ProtoReflect() protoreflect.Message {
	mi := &file_proto_ethaudit_proto_msgTypes[1]
	if protoimpl.UnsafeEnabled && x != nil {
		ms := protoimpl.X.MessageStateOf(protoimpl.Pointer(x))
		if ms.LoadMessageInfo() == nil {
			ms.StoreMessageInfo(mi)
		}
		return ms
	}
	return mi.MessageOf(x)
}

// Deprecated: Use Status.ProtoReflect.Descriptor instead.
func (*Status) Descriptor() ([]byte, []int) {
	return file_proto_ethaudit_proto_rawDescGZIP(), []int{1}
}

func (x *Status) GetCode() StatusCode {
	if x != nil {
		return x.Code
	}
	return StatusCode_Unknwon
}

var File_proto_ethaudit_proto protoreflect.FileDescriptor

var file_proto_ethaudit_proto_rawDesc = []byte{
	0x0a, 0x14, 0x70, 0x72, 0x6f, 0x74, 0x6f, 0x2f, 0x65, 0x74, 0x68, 0x61, 0x75, 0x64, 0x69, 0x74,
	0x2e, 0x70, 0x72, 0x6f, 0x74, 0x6f, 0x12, 0x08, 0x65, 0x74, 0x68, 0x61, 0x75, 0x64, 0x69, 0x74,
	0x22, 0xaf, 0x03, 0x0a, 0x09, 0x42, 0x6c, 0x6f, 0x63, 0x6b, 0x49, 0x6e, 0x66, 0x6f, 0x12, 0x1e,
	0x0a, 0x0a, 0x50, 0x61, 0x72, 0x65, 0x6e, 0x74, 0x48, 0x61, 0x73, 0x68, 0x18, 0x01, 0x20, 0x01,
	0x28, 0x0c, 0x52, 0x0a, 0x50, 0x61, 0x72, 0x65, 0x6e, 0x74, 0x48, 0x61, 0x73, 0x68, 0x12, 0x1c,
	0x0a, 0x09, 0x55, 0x6e, 0x63, 0x6c, 0x65, 0x48, 0x61, 0x73, 0x68, 0x18, 0x02, 0x20, 0x01, 0x28,
	0x0c, 0x52, 0x09, 0x55, 0x6e, 0x63, 0x6c, 0x65, 0x48, 0x61, 0x73, 0x68, 0x12, 0x1a, 0x0a, 0x08,
	0x43, 0x6f, 0x69, 0x6e, 0x62, 0x61, 0x73, 0x65, 0x18, 0x03, 0x20, 0x01, 0x28, 0x0c, 0x52, 0x08,
	0x43, 0x6f, 0x69, 0x6e, 0x62, 0x61, 0x73, 0x65, 0x12, 0x12, 0x0a, 0x04, 0x52, 0x6f, 0x6f, 0x74,
	0x18, 0x04, 0x20, 0x01, 0x28, 0x0c, 0x52, 0x04, 0x52, 0x6f, 0x6f, 0x74, 0x12, 0x16, 0x0a, 0x06,
	0x54, 0x78, 0x48, 0x61, 0x73, 0x68, 0x18, 0x05, 0x20, 0x01, 0x28, 0x0c, 0x52, 0x06, 0x54, 0x78,
	0x48, 0x61, 0x73, 0x68, 0x12, 0x20, 0x0a, 0x0b, 0x52, 0x65, 0x63, 0x65, 0x69, 0x70, 0x74, 0x48,
	0x61, 0x73, 0x68, 0x18, 0x06, 0x20, 0x01, 0x28, 0x0c, 0x52, 0x0b, 0x52, 0x65, 0x63, 0x65, 0x69,
	0x70, 0x74, 0x48, 0x61, 0x73, 0x68, 0x12, 0x14, 0x0a, 0x05, 0x42, 0x6c, 0x6f, 0x6f, 0x6d, 0x18,
	0x07, 0x20, 0x01, 0x28, 0x0c, 0x52, 0x05, 0x42, 0x6c, 0x6f, 0x6f, 0x6d, 0x12, 0x1e, 0x0a, 0x0a,
	0x44, 0x69, 0x66, 0x66, 0x69, 0x63, 0x75, 0x6c, 0x74, 0x79, 0x18, 0x08, 0x20, 0x01, 0x28, 0x04,
	0x52, 0x0a, 0x44, 0x69, 0x66, 0x66, 0x69, 0x63, 0x75, 0x6c, 0x74, 0x79, 0x12, 0x16, 0x0a, 0x06,
	0x4e, 0x75, 0x6d, 0x62, 0x65, 0x72, 0x18, 0x09, 0x20, 0x01, 0x28, 0x04, 0x52, 0x06, 0x4e, 0x75,
	0x6d, 0x62, 0x65, 0x72, 0x12, 0x1a, 0x0a, 0x08, 0x47, 0x61, 0x73, 0x4c, 0x69, 0x6d, 0x69, 0x74,
	0x18, 0x0a, 0x20, 0x01, 0x28, 0x04, 0x52, 0x08, 0x47, 0x61, 0x73, 0x4c, 0x69, 0x6d, 0x69, 0x74,
	0x12, 0x18, 0x0a, 0x07, 0x47, 0x61, 0x73, 0x55, 0x73, 0x65, 0x64, 0x18, 0x0b, 0x20, 0x01, 0x28,
	0x04, 0x52, 0x07, 0x47, 0x61, 0x73, 0x55, 0x73, 0x65, 0x64, 0x12, 0x12, 0x0a, 0x04, 0x54, 0x69,
	0x6d, 0x65, 0x18, 0x0c, 0x20, 0x01, 0x28, 0x04, 0x52, 0x04, 0x54, 0x69, 0x6d, 0x65, 0x12, 0x14,
	0x0a, 0x05, 0x45, 0x78, 0x74, 0x72, 0x61, 0x18, 0x0d, 0x20, 0x01, 0x28, 0x0c, 0x52, 0x05, 0x45,
	0x78, 0x74, 0x72, 0x61, 0x12, 0x1c, 0x0a, 0x09, 0x4d, 0x69, 0x78, 0x44, 0x69, 0x67, 0x65, 0x73,
	0x74, 0x18, 0x0e, 0x20, 0x01, 0x28, 0x0c, 0x52, 0x09, 0x4d, 0x69, 0x78, 0x44, 0x69, 0x67, 0x65,
	0x73, 0x74, 0x12, 0x14, 0x0a, 0x05, 0x4e, 0x6f, 0x6e, 0x63, 0x65, 0x18, 0x0f, 0x20, 0x01, 0x28,
	0x04, 0x52, 0x05, 0x4e, 0x6f, 0x6e, 0x63, 0x65, 0x12, 0x18, 0x0a, 0x07, 0x42, 0x61, 0x73, 0x65,
	0x46, 0x65, 0x65, 0x18, 0x10, 0x20, 0x01, 0x28, 0x04, 0x52, 0x07, 0x42, 0x61, 0x73, 0x65, 0x46,
	0x65, 0x65, 0x22, 0x32, 0x0a, 0x06, 0x53, 0x74, 0x61, 0x74, 0x75, 0x73, 0x12, 0x28, 0x0a, 0x04,
	0x43, 0x6f, 0x64, 0x65, 0x18, 0x01, 0x20, 0x01, 0x28, 0x0e, 0x32, 0x14, 0x2e, 0x65, 0x74, 0x68,
	0x61, 0x75, 0x64, 0x69, 0x74, 0x2e, 0x53, 0x74, 0x61, 0x74, 0x75, 0x73, 0x43, 0x6f, 0x64, 0x65,
	0x52, 0x04, 0x43, 0x6f, 0x64, 0x65, 0x2a, 0x2d, 0x0a, 0x0a, 0x53, 0x74, 0x61, 0x74, 0x75, 0x73,
	0x43, 0x6f, 0x64, 0x65, 0x12, 0x0b, 0x0a, 0x07, 0x55, 0x6e, 0x6b, 0x6e, 0x77, 0x6f, 0x6e, 0x10,
	0x00, 0x12, 0x06, 0x0a, 0x02, 0x4f, 0x6b, 0x10, 0x01, 0x12, 0x0a, 0x0a, 0x06, 0x46, 0x61, 0x69,
	0x6c, 0x65, 0x64, 0x10, 0x02, 0x42, 0x3a, 0x5a, 0x38, 0x67, 0x69, 0x74, 0x68, 0x75, 0x62, 0x2e,
	0x63, 0x6f, 0x6d, 0x2f, 0x64, 0x6f, 0x63, 0x62, 0x75, 0x6c, 0x6c, 0x2f, 0x69, 0x6e, 0x6c, 0x61,
	0x62, 0x2d, 0x65, 0x74, 0x68, 0x65, 0x72, 0x65, 0x75, 0x6d, 0x2d, 0x68, 0x70, 0x32, 0x70, 0x2f,
	0x61, 0x75, 0x64, 0x69, 0x74, 0x63, 0x68, 0x61, 0x69, 0x6e, 0x2f, 0x70, 0x72, 0x6f, 0x74, 0x6f,
	0x2f, 0x62, 0x06, 0x70, 0x72, 0x6f, 0x74, 0x6f, 0x33,
}

var (
	file_proto_ethaudit_proto_rawDescOnce sync.Once
	file_proto_ethaudit_proto_rawDescData = file_proto_ethaudit_proto_rawDesc
)

func file_proto_ethaudit_proto_rawDescGZIP() []byte {
	file_proto_ethaudit_proto_rawDescOnce.Do(func() {
		file_proto_ethaudit_proto_rawDescData = protoimpl.X.CompressGZIP(file_proto_ethaudit_proto_rawDescData)
	})
	return file_proto_ethaudit_proto_rawDescData
}

var file_proto_ethaudit_proto_enumTypes = make([]protoimpl.EnumInfo, 1)
var file_proto_ethaudit_proto_msgTypes = make([]protoimpl.MessageInfo, 2)
var file_proto_ethaudit_proto_goTypes = []interface{}{
	(StatusCode)(0),   // 0: ethaudit.StatusCode
	(*BlockInfo)(nil), // 1: ethaudit.BlockInfo
	(*Status)(nil),    // 2: ethaudit.Status
}
var file_proto_ethaudit_proto_depIdxs = []int32{
	0, // 0: ethaudit.Status.Code:type_name -> ethaudit.StatusCode
	1, // [1:1] is the sub-list for method output_type
	1, // [1:1] is the sub-list for method input_type
	1, // [1:1] is the sub-list for extension type_name
	1, // [1:1] is the sub-list for extension extendee
	0, // [0:1] is the sub-list for field type_name
}

func init() { file_proto_ethaudit_proto_init() }
func file_proto_ethaudit_proto_init() {
	if File_proto_ethaudit_proto != nil {
		return
	}
	if !protoimpl.UnsafeEnabled {
		file_proto_ethaudit_proto_msgTypes[0].Exporter = func(v interface{}, i int) interface{} {
			switch v := v.(*BlockInfo); i {
			case 0:
				return &v.state
			case 1:
				return &v.sizeCache
			case 2:
				return &v.unknownFields
			default:
				return nil
			}
		}
		file_proto_ethaudit_proto_msgTypes[1].Exporter = func(v interface{}, i int) interface{} {
			switch v := v.(*Status); i {
			case 0:
				return &v.state
			case 1:
				return &v.sizeCache
			case 2:
				return &v.unknownFields
			default:
				return nil
			}
		}
	}
	type x struct{}
	out := protoimpl.TypeBuilder{
		File: protoimpl.DescBuilder{
			GoPackagePath: reflect.TypeOf(x{}).PkgPath(),
			RawDescriptor: file_proto_ethaudit_proto_rawDesc,
			NumEnums:      1,
			NumMessages:   2,
			NumExtensions: 0,
			NumServices:   0,
		},
		GoTypes:           file_proto_ethaudit_proto_goTypes,
		DependencyIndexes: file_proto_ethaudit_proto_depIdxs,
		EnumInfos:         file_proto_ethaudit_proto_enumTypes,
		MessageInfos:      file_proto_ethaudit_proto_msgTypes,
	}.Build()
	File_proto_ethaudit_proto = out.File
	file_proto_ethaudit_proto_rawDesc = nil
	file_proto_ethaudit_proto_goTypes = nil
	file_proto_ethaudit_proto_depIdxs = nil
}


extractor/main.go
package main

import (
	"encoding/hex"
	"fmt"
	"io"
	"net"

	extractor "github.com/docbull/ethereum-block-extractor/proto"
	proto "github.com/golang/protobuf/proto"
)

func ReceiveBlockInfo(conn net.Conn) {
	data := make([]byte, 0)
	buf := make([]byte, 1024)
	length := 0

	for {
		n, err := conn.Read(buf)
		if err != nil {
			if err != io.EOF {
				fmt.Println(err)
				return
			}
			break
		}

		data = append(data, buf[:n]...)
		length += n
	}

	block := &extractor.BlockInfo{}
	err := proto.Unmarshal(data[:length], block)
	if err != nil {
		fmt.Println(err)
		conn.Write([]byte("error"))
		return
	}

	fmt.Println("difficulty:", block.Difficulty)
	fmt.Println("extraData:", hex.EncodeToString(block.Extra))
	fmt.Println("gasLimit:", block.GasLimit)
	fmt.Println("gasUsed:", block.GasUsed)
	fmt.Println("hash:", hex.EncodeToString(block.TxHash))
	fmt.Println("logsBloom:", hex.EncodeToString(block.Bloom))
	fmt.Println("miner:", hex.EncodeToString(block.Coinbase))
	fmt.Println("mixHash:", hex.EncodeToString(block.MixDigest))
	fmt.Println("nonce:", block.Nonce)
	fmt.Println("number:", block.Number)
	fmt.Println("parentHash:", hex.EncodeToString(block.ParentHash))
	fmt.Println("receiptsRoot:", hex.EncodeToString(block.ReceiptHash))
	fmt.Println("sha3Uncles:", hex.EncodeToString(block.UncleHash))
	fmt.Println("stateRoot:", hex.EncodeToString(block.Root))
	fmt.Println("timestamp:", block.Time)
	fmt.Println("transactionRoot:", hex.EncodeToString(block.TxHash))
}

func Start2Listen() {
	lis, err := net.Listen("tcp", ":4242")
	if err != nil {
		fmt.Println(err)
		return
	}
	defer lis.Close()

	fmt.Println("🚀 Ethereum Block Extractor ...")
	for {
		conn, err := lis.Accept()
		fmt.Println("🎁 New Block Arrived!")
		if err != nil {
			fmt.Println(err)
			return
		}

		go ReceiveBlockInfo(conn)
	}
}

func main() {
	Start2Listen()
}
